docs/madr/decisions/014-mesh-circuit-breaker-policy.md

```diff
@@ -0,0 +1,198 @@
+debugging process, may look only at stats prefixed with `cluster.<name>.circuit_breakers.`, but according to the
+behaviour of our policy, if the data plane was ejected because of one of the faults defined in `detectors` section,
+stats which contain interesting us data will be prefixed with `cluster.<name>.outlier_detection.`.
+
+## Considered Options
+
+* Create two separate policies: `MeshCircuitBreaker` and `MeshOutlierDetector`
+* Create single `MeshCircuitBreaker` policy which would map 1:1 current policy
+* Create single `MeshCircuitBreaker` with configuration divided
+  to `connectionPool` ([envoy's circuit breaker](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/circuit_breaking))
+  and `outlierDetection` ([envoy's outlier detection](https://gstwww.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/outlier))
+
+## Decision Outcome
+
+Chosen option: Create single `MeshCircuitBreaker` with configuration divided into two sections.
+
+## Positive Consequences
+
+* Having two separate policies, each with explicit configurations makes it easier to look for appropriate Envoy stats
+  and XDP configuration values.
+* By combining two abstractions, which exist in Envoy, and calling them both combined with a name of one of these is
+  confusing, especially when you are trying to debug some issue without deep understanding of internals of our service
+  mesh.
+* Default `MeshCircuitBreaker` policy is sufficient to replicate behaviour of current version of this policy - there
+  won't be a need to include a default `MeshOutlierDetector` policy
+
+## Negative Consequences
+
+* `MeshCircuitBreaker` will be a policy with very small amount of configuration available, which may introduce some
+  clutter (at some point there may be so many policies, that it will be much more difficult to understand all of them)
+
+## Solution
+
+By dividing the configuration section into two separate, logical sections: `connectionPool` and `outlierDetection` it's
+easier to differentiate underlying envoy features without the need to split policy into two separate ones.
+
+Additional changes to the original `CircuitBreaker` policy consists of:
+
+- flattening `detectors.*` options with only single (`.consecutive`) property
+- changing word `Errors` to `Failures`
+- adding `origin` suffix to the local failures detector
+
+It means `totalErrors.consecutive`, `gatewayErrors.consecutive` and `localErrors.consecutive`
+becomes: `consecutiveTotalFailures`, `consecutiveGatewayFailures` and `consecutiveLocalOriginFailures` respectively.
+
+### Current configuration
+
+Below is the sample `MeshCircuitBreaker` configuration
+
+```yaml
+conf:
+  interval: 1s
+  baseEjectionTime: 30s
+  maxEjectionPercent: 20
+  splitExternalAndLocalErrors: false
+  thresholds:
+    maxConnections: 2
+    maxPendingRequests: 2
+    maxRequests: 2
+    maxRetries: 2
```

Comment:
What is max retries and how does it relate to Retry policy?

Reply:
I will try to dig into it a little bit further

Reply:
Ok, I have confirmation. If there are more connections being retried in parallel than specified in this configuration, the circuit breaker will be tripped, which means if the next connection would like to be retried according to the `Retry` policy, circuit breaker won't allow it and `upstream_rq_retry_overflow` stat will be updated.